src/Data/Tree/Binary/Zipper.agda

```diff
@@ -0,0 +1,105 @@
+    A : Set a
+    B : Set b
+    C : Set c
+    D : Set d
+
+data Crumb (A : Set a) (B : Set b) : Set (a ⊔ b) where
+  leftBranch : A → Tree A B → Crumb A B
+  rightBranch : A → Tree A B → Crumb A B
+
+record Zipper (A : Set a) (B : Set b) : Set (a ⊔ b) where
+  constructor mkZipper
+  field
+    context : List (Crumb A B)
+    focus : Tree A B
+
+open Zipper public
+
+-- Fundamental operations of a Zipper: Moving around
+------------------------------------------------------------------------
+
+up : Zipper A B → Maybe (Zipper A B)
+up (mkZipper [] foc) = nothing
+up (mkZipper (leftBranch m l ∷ ctx) foc) = just $ mkZipper ctx (node l m foc)
+up (mkZipper (rightBranch m r ∷ ctx) foc) = just $ mkZipper ctx (node foc m r)
+
+left : Zipper A B → Maybe (Zipper A B)
+left (mkZipper ctx (leaf x)) = nothing
+left (mkZipper ctx (node l m r)) = just $ mkZipper (rightBranch m r ∷ ctx) l
+
+right : Zipper A B → Maybe (Zipper A B)
+right (mkZipper ctx (leaf x)) = nothing
+right (mkZipper ctx (node l m r)) = just $ mkZipper (leftBranch m l ∷ ctx) r
+
+-- To and from trees
+------------------------------------------------------------------------
+
+toTreeHelper : List (Crumb A B) → Tree A B → Tree A B
+toTreeHelper [] t = t
+toTreeHelper (leftBranch m l ∷ xs) t = toTreeHelper xs (node l m t)
+toTreeHelper (rightBranch m r ∷ xs) t = toTreeHelper xs (node t m r)
+
+toTree : Zipper A B → Tree A B
+toTree (mkZipper ctx foc) = toTreeHelper ctx foc
+
+fromTree : Tree A B → Zipper A B
+fromTree = mkZipper []
+
+-- Tree-like operations
+------------------------------------------------------------------------
+
+getTree : Crumb A B → Tree A B
+getTree (leftBranch m x) = x
+getTree (rightBranch m x) = x
+
+mapCrumb : (A → C) → (B → D) → Crumb A B → Crumb C D
+mapCrumb f g (leftBranch m x) = leftBranch (f m) (BT.map f g x)
+mapCrumb f g (rightBranch m x) = rightBranch (f m) (BT.map f g x)
+
+#nodes : Zipper A B → ℕ
+#nodes (mkZipper ctx foc) = BT.#nodes foc + sum (List.map (suc ∘ BT.#nodes ∘ getTree) ctx)
+
+#leaves : Zipper A B → ℕ
+#leaves (mkZipper ctx foc) = BT.#leaves foc + sum (List.map (BT.#leaves ∘ getTree) ctx)
+
+map : (A → C) → (B → D) → Zipper A B → Zipper C D
+map f g (mkZipper ctx foc) = mkZipper (List.map (mapCrumb f g) ctx) (BT.map f g foc)
+
+foldr : (C → A → C → C) → (B → C) → Zipper A B → C
+foldr {C = C} {A = A} {B = B} f g (mkZipper ctx foc) = List.foldl step (BT.foldr f g foc) ctx
+  where
+    step : C → Crumb A B → C
+    step val (leftBranch m x) = f (BT.foldr f g x) m val
+    step val (rightBranch m x) = f val m (BT.foldr f g x)
+
+-- Attach nodes to the top most part of the zipper
+------------------------------------------------------------------------
+
+infixr 5 _⟪_⟫ˡ_
+infixl 5 _⟪_⟫ʳ_
+
+_⟪_⟫ˡ_ : Tree A B → A → Zipper A B → Zipper A B
+l ⟪ m ⟫ˡ mkZipper ctx foc = mkZipper (ctx ++ [ (leftBranch m l) ]) foc
+
+_⟪_⟫ʳ_ : Zipper A B → A → Tree A B → Zipper A B
+mkZipper ctx foc ⟪ m ⟫ʳ r = mkZipper (ctx ++ [ (rightBranch m r) ]) foc
```

Comment:
These operations are costly. They should be generalised to a full
`Zipper A B -> List (Crumb A B) -> Zipper A B` so that we can pour
a whole one-hole context on top of the current one without iterating
over `_⟪_⟫ˡ_` (it would be quadratic instead of linear). We can then
have them as a special case.